ddl/reorg.go

```diff
@@ -318,62 +348,50 @@ func getReorgInfo(d *ddlCtx, t *meta.Meta, job *model.Job, tbl table.Table) (*re
 			return nil, errInvalidStoreVer.Gen("invalid storage current version %d", ver.Ver)
 		}
 
-		// Get the first handle of this table.
-		err = iterateSnapshotRows(d.store, tbl, ver.Ver, math.MinInt64,
-			func(h int64, rowKey kv.Key, rawRecord []byte) (bool, error) {
-				info.StartHandle = h
-				return false, nil
-			})
-		if err != nil {
-			return info, errors.Trace(err)
+		tblInfo := tbl.Meta()
+		pid := tblInfo.ID
+		if pi := tblInfo.GetPartitionInfo(); pi != nil {
+			pid = pi.Definitions[0].ID
 		}
+		start, end, err1 := getTableRange(d, tblInfo, pid, ver.Ver)
+		if err1 != nil {
+			return nil, errors.Trace(err1)
+		}
+		log.Infof("[ddl-reorg] job %v get partition %d range [%d %d]", job.ID, pid, info.StartHandle, info.EndHandle)
 
 		// gofail: var errorUpdateReorgHandle bool
 		// if errorUpdateReorgHandle && !gofailOnceGuard {
 		//  // only return error once.
 		//	gofailOnceGuard = true
 		// 	return info, errors.New("occur an error when update reorg handle.")
 		// }
-		err = t.UpdateDDLReorgHandle(job, info.StartHandle)
+		err = t.UpdateDDLReorgHandle(job, start, end, pid)
 		if err != nil {
 			return info, errors.Trace(err)
 		}
+		// Update info should after data persistent.
+		info.StartHandle = start
+		info.EndHandle = end
+		info.PartitionID = pid
 
 		job.SnapshotVer = ver.Ver
 	} else {
-		info.StartHandle, err = t.GetDDLReorgHandle(job)
+		start, end, pid, err := t.GetDDLReorgHandle(job)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		info.StartHandle = start
+		info.EndHandle = end
+		info.PartitionID = pid
```

Comment:
Could we put line380-385 after line386? Then we can remove line369-375. And we can put line331 here.